proto/kvrpcpb.proto

```diff
@@ -1,61 +1,160 @@
 syntax = "proto2";
 package kvrpcpb;
+import "metapb.proto";
 
 enum MessageType {
     CmdGet = 1;
     CmdScan = 2;
     CmdPrewrite = 3;
     CmdCommit = 4;
+    CmdCleanup = 5;
+    // Below types both use for Get failed. If Get failed, it may be locked.
+    // So it tries to clean primary lock(CmdCleanup), and then server will return
+    // either committed or rolled back. Finally, client will commit/rollback
+    // primary lock and THEN Get again.
+    CmdRollbackThenGet = 6;
+    CmdCommitThenGet = 7;
 }
+
+message ResultType {
+    enum Type {
+        Ok = 1;
+        Retryable = 2;
+        Locked = 3;
+        // Known result type add here.
+        Other = 9;
+    }
+    optional Type type = 1 [default = Ok];
+    optional string msg = 2;
+}
+
+message KeyAddress {
+    optional bytes key = 1;
+    optional uint64 region_id = 2;
+    optional metapb.Peer peer = 3;
+}
+
 message KvPair {
-    required bytes key = 1;
-    required bytes value = 2;
+    optional KeyAddress key_address = 1;
+    optional bytes value = 2;
 }
 
 message CmdGetRequest {
-    required bytes key = 1;
-    required uint64 version = 2;
+    optional KeyAddress key_address = 1;
+    optional uint64 version = 2;
 }
+
 message CmdGetResponse {
-    required bool ok = 1;
+    optional ResultType res_type = 1;
     optional bytes value = 2;
+    optional bytes primary_lock = 3;
+    optional uint64 lock_version = 4;
 }
+
 message CmdScanRequest {
-    required bytes key = 1;
-    required uint32 limit = 2;
-    required uint64 version = 3;
+    optional KeyAddress key_address = 1;
+    optional uint32 limit = 2;
+    optional uint64 version = 3;
 }
+
+message Item {
+    optional ResultType res_type = 1;
+    optional bytes key = 2;
+    optional bytes value = 3;
+    // primary_lock_key
+    optional bytes primary_lock = 4;
+    optional uint64 lock_version = 5;
+}
+
 message CmdScanResponse {
-    required bool ok = 1;
-    repeated KvPair results = 2;
+    // ok if !ok then retry.
+    optional bool ok = 1;
+    repeated Item results = 2;
 }
+
 message CmdPrewriteRequest {
     repeated KvPair puts = 1;
-    repeated bytes dels = 2;
-    repeated bytes locks = 3;
-    required uint64 start_version = 4;
+    repeated KeyAddress dels = 2;
+    repeated KeyAddress locks = 3;
+    // primary_lock_key_address
+    optional KeyAddress primary_address = 4;
```

Comment:
Server will not use primary lock as a key to query raft engine. So `bytes` is enough.